jgrapht-core/src/main/java/org/jgrapht/graph/DirectedSubgraph.java

```diff
@@ -51,79 +51,65 @@ public DirectedSubgraph(DirectedGraph<V, E> base, Set<V> vertexSubset, Set<E> ed
     }
 
     /**
-     * @see DirectedGraph#inDegreeOf(Object)
+     * Creates a new directed induced subgraph.
+     *
+     * @param base the base (backing) graph on which the subgraph will be based.
+     * @param vertexSubset vertices to include in the subgraph. If <code>
+     * null</code> then all vertices are included.
      */
-    @Override
-    public int inDegreeOf(V vertex)
+    public DirectedSubgraph(DirectedGraph<V, E> base, Set<V> vertexSubset)
     {
-        assertVertexExist(vertex);
-
-        int degree = 0;
+        this(base, vertexSubset, null);
+    }
 
-        for (E e : getBase().incomingEdgesOf(vertex)) {
-            if (containsEdge(e)) {
-                degree++;
-            }
-        }
+    /**
+     * Creates a new directed induced subgraph with all vertices included.
+     *
+     * @param base the base (backing) graph on which the subgraph will be based.
+     */
+    public DirectedSubgraph(DirectedGraph<V, E> base)
+    {
+        this(base, null, null);
+    }
 
-        return degree;
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int inDegreeOf(V vertex)
+    {
+        return incomingEdgesOf(vertex).size();
     }
 
     /**
-     * @see DirectedGraph#incomingEdgesOf(Object)
+     * {@inheritDoc}
      */
     @Override
     public Set<E> incomingEdgesOf(V vertex)
     {
         assertVertexExist(vertex);
 
-        Set<E> edges = new ArrayUnenforcedSet<>();
-
-        for (E e : getBase().incomingEdgesOf(vertex)) {
-            if (containsEdge(e)) {
-                edges.add(e);
-            }
-        }
-
-        return edges;
+        return new FilteredSet<>(base.incomingEdgesOf(vertex), e -> edgeSet.contains(e));
```

Comment:
I somehow wanted to make the behavior more "live", i.e. to reflect the changes that may happen in the base graph.